In [5]:
# 0. 모듈 불러오기

import pandas as pd
import numpy as np

In [3]:
# 1. 데이터 불러오기

train_data = pd.read_json('data/train.json', encoding='UTF-8', typ = 'frame')
song_meta = pd.read_json('data/song_meta.json', encoding='UTF-8', typ = 'frame')
genre_gn_all = pd.read_json('data/genre_gn_all.json',encoding='UTF-8', typ = 'series')
genre_gn_all = pd.DataFrame(genre_gn_all, columns = ['gnr_name']).reset_index().rename(columns = {'index' : 'gnr_code'})

### 데이터를 확인해보자!

In [4]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115071 entries, 0 to 115070
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   tags          115071 non-null  object
 1   id            115071 non-null  int64 
 2   plylst_title  115071 non-null  object
 3   songs         115071 non-null  object
 4   like_cnt      115071 non-null  int64 
 5   updt_date     115071 non-null  object
dtypes: int64(2), object(4)
memory usage: 5.3+ MB


In [5]:
song_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 707989 entries, 0 to 707988
Data columns (total 9 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   song_gn_dtl_gnr_basket  707989 non-null  object
 1   issue_date              707989 non-null  int64 
 2   album_name              707985 non-null  object
 3   album_id                707989 non-null  int64 
 4   artist_id_basket        707989 non-null  object
 5   song_name               707989 non-null  object
 6   song_gn_gnr_basket      707989 non-null  object
 7   artist_name_basket      707989 non-null  object
 8   id                      707989 non-null  int64 
dtypes: int64(3), object(6)
memory usage: 48.6+ MB


In [6]:
genre_gn_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 254 entries, 0 to 253
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   gnr_code  254 non-null    object
 1   gnr_name  254 non-null    object
dtypes: object(2)
memory usage: 4.1+ KB


#### train_data 크기 : 115071 (약 11만)
#### song 크기 : 707989 (약 70만)
#### genre_gn_all : 254

In [6]:
train_data.head()

tags      id  \
0                                                [락]   61281   
1                                           [추억, 회상]   10532   
2                                          [까페, 잔잔한]   76951   
3  [연말, 눈오는날, 캐럴, 분위기, 따듯한, 크리스마스캐럴, 겨울노래, 크리스마스,...  147456   
4                                               [댄스]   27616   

            plylst_title                                              songs  \
0                여행같은 음악  [525514, 129701, 383374, 562083, 297861, 13954...   
1                요즘 너 말야  [432406, 675945, 497066, 120377, 389529, 24427...   
2  편하게, 잔잔하게 들을 수 있는 곡.-  [83116, 276692, 166267, 186301, 354465, 256598...   
3  크리스마스 분위기에 흠뻑 취하고 싶을때  [394031, 195524, 540149, 287984, 440773, 10033...   
4               추억의 노래 ㅋ  [159327, 553610, 5130, 645103, 294435, 100657,...   

   like_cnt                updt_date  
0        71  2013-12-19 18:36:19.000  
1         1  2014-12-02 16:19:42.000  
2        17  2017-08-28 07:09:34.000  
3        33  2019-12-05 15:15:18.000  
4         9  2011-10-25 13:54:56.000

In [11]:
song_meta['id'].describe()

count    707989.000000
mean     353994.000000
std      204378.964203
min           0.000000
25%      176997.000000
50%      353994.000000
75%      530991.000000
max      707988.000000
Name: id, dtype: float64

In [12]:
train_data['id'].describe()

count    115071.000000
mean      76728.259501
std       44313.134619
min           1.000000
25%       38277.500000
50%       76732.000000
75%      115061.500000
max      153428.000000
Name: id, dtype: float64

In [7]:
genre_gn_all

gnr_code gnr_name
0     GN0100      발라드
1     GN0101   세부장르전체
2     GN0102      '80
3     GN0103      '90
4     GN0104      '00
..       ...      ...
249   GN2900      뮤지컬
250   GN2901   세부장르전체
251   GN2902    국내뮤지컬
252   GN2903    국외뮤지컬
253   GN3000    크리스마스

[254 rows x 2 columns]

## processed 된 Data load
#### carol들로 이루어진 플레이리스트 + carol 들

In [6]:
import import_ipynb
from util import load_json

train_playlist = load_json("./data/processedData/playlist_train.json")
val_playlist = load_json("./data/processedData/playlist_val.json")
test_playlist = load_json("./data/processedData/playlist_test.json")
carol_data = load_json("./data/processedData/carol_data.json")
train_playlist = pd.DataFrame(eval(train_playlist))
val_playlist = pd.DataFrame(eval(val_playlist))
test_playlist = pd.DataFrame(eval(test_playlist))
carol_data = pd.DataFrame(eval(carol_data))

In [19]:
print(f'Num of train_data: {len(train_playlist)}')
train_playlist.head()

Num of train_data: 1030


tags      id  \
0                            [JPOP, 매장음악, 크리스마스, 일본]   24882   
1   [연말, 겨울감성, 연주곡, 눈, 성탄절, 크리스마스, 설렘, 캐롤, 자장가, 캐롤송]  124080   
2                                       [겨울, 설렘, 사랑]   15936   
3                                           [겨울, 까페]   24535   
4  [겨울, 눈오는날, 파티, 카페, 매장, 드라이브, 여행, 크리스마스, 캐롤, 2016]   44592   

        plylst_title                                              songs  \
0   jpop 크리스마스 캐롤 등등  [171751, 25810, 653717, 631818, 486608, 237975...   
1  설레는 크리스마스 캐롤연주곡:)  [220071, 28532, 159001, 329291, 193160, 417231...   
2       ☃메리해피✰크리스마스☃  [404382, 456575, 405143, 478191, 245321, 21731...   
3        크리스마스 카페 음악  [357064, 137969, 113394, 628504, 205686, 70626...   
4   크리스마스 캐롤 여행을 떠나요  [355931, 164891, 261338, 52934, 695032, 456704...   

   like_cnt                updt_date  
0        50  2017-12-17 02:33:47.000  
1         8  2018-11-23 23:49:10.000  
2         0  2014-12-22 12:16:33.000  
3       145  2013-12-04 12:46:13.000  
4        15  2016-12-06 01:15:51.000

#### 우리가 사용할 train_playlist의 songs의 크기 정보!

In [16]:
from collections import Counter

train_copy = train_playlist.copy()

train_copy['songs_id'] = train_copy['songs'].map(lambda x: [song_id_sid.get(s) for s in x if song_id_sid.get(s) != None])

plylst_song = train_copy['songs']
song_counter = Counter([sg for sgs in plylst_song for sg in sgs])
song_dict = {x: song_counter[x] for x in song_counter}

song_id_sid = dict()
song_sid_id = dict()
for i, t in enumerate(song_dict):
  song_id_sid[t] = i
  song_sid_id[i] = t

train_copy.loc[:,'num_songs'] = train_copy['songs_id'].map(len)

train_copy['num_songs'].describe()

count    1030.000000
mean       34.850485
std        36.672492
min         6.000000
25%        10.000000
50%        21.000000
75%        42.000000
max       200.000000
Name: num_songs, dtype: float64

In [22]:
print(f'Num of carol_data: {len(carol_data)}')
carol_data.head()

Num of carol_data: 15050


song_gn_dtl_gnr_basket  issue_date                          album_name  \
0  [GN1601, GN1615, GN1610]    20171006      Christmas at Westminster Abbey   
1  [GN1701, GN1703, GN1702]    20121109  Christmas (Deluxe Special Edition)   
2                  [GN0101]    20190408                                있어줄래   
3          [GN0105, GN0101]    20121204                       WINTER POETRY   
4          [GN1304, GN1301]    20110810                        Simple Steps   

   album_id         artist_id_basket  \
0  10104201  [345328, 36652, 347399]   
1   2167838                  [35370]   
2  10271322                 [500421]   
3   2169041                   [1741]   
4   2058451                 [614091]   

                                           song_name song_gn_gnr_basket  \
0  Anonymous: Up! Awake! From Highest Steeple - A...           [GN1600]   
1          Mis Deseos \/ Feliz Navidad (with Thalia)           [GN1700]   
2                                               있어줄래           [GN0100]   
3                                       너에게 전하지 못한 말           [GN0100]   
4                                      Rollercoaster           [GN1300]   

                                  artist_name_basket  id  
0  [Choir Of Westminster Abbey, Simon Preston, Ch...  23  
1                                    [Michael Buble]  41  
2                                             [길구봉구]  54  
3                                              [신혜성]  70  
4                                          [Sam Ock]  71

In [20]:
print(f'Num of validation_data: {len(val_playlist)}')
val_playlist.head()

Num of validation_data: 258


tags      id             plylst_title  \
0                      [연인, 겨울, 크리스마스]   70701      크리스마스에 들으면 기분좋아지는노래   
1       [달콤, 따뜻, 달달, 핑크핑크, 케이크, 크리스마스]  120113   크리스마스 케이크 같은 달달 따뜻한 음악   
2  [달달한노래, 메리크리스마스, 연인, 행복, 크리스마스, 사랑]  146012  크리스마스 연인들을 위한 노래 요기있어요!   
3                            [휴식, 발라드]  113900                 크리스마스125   
4                            [휴식, 잔잔한]   28001                  크리스마스54   

                                               songs  like_cnt  \
0  [397351, 410185, 595353, 187149, 326204, 17496...         2   
1  [613003, 95485, 581552, 686406, 692454, 432211...         3   
2  [614187, 572996, 404950, 563199, 600640, 24840...         4   
3  [405485, 381653, 566122, 546313, 158539, 68697...         0   
4  [405485, 381653, 566122, 546313, 158539, 68697...         0   

                 updt_date  
0  2017-12-26 09:59:58.000  
1  2016-12-06 16:38:56.000  
2  2016-12-24 13:51:31.000  
3  2019-11-19 11:37:06.000  
4  2019-11-19 11:06:50.000

In [21]:
print(f'Num of test_data: {len(test_playlist)}')
test_playlist.head()

Num of test_data: 322


tags      id        plylst_title  \
0   [기분전환, 드라이브]  115037            크리스마스387   
1    [드라이브, 발라드]   22805            크리스마스204   
2  [재즈, 캐롤, 한여름]   28570  한여름의 크리스마스 재즈 캐롤?!   
3           [겨울]   57791        [크리스마스 캐롤모음]   
4    [기분전환, 잔잔한]   25370            크리스마스138   

                                               songs  like_cnt  \
0  [405485, 381653, 566122, 546313, 158539, 68697...         0   
1  [405485, 381653, 566122, 546313, 158539, 68697...         0   
2  [262793, 226323, 40371, 50235, 523943, 377551,...         1   
3  [226670, 404382, 653711, 370785, 662376, 64336...        82   
4  [405485, 381653, 566122, 546313, 158539, 68697...         0   

                 updt_date  
0  2019-11-19 15:10:15.000  
1  2019-11-19 13:22:15.000  
2  2016-07-14 10:19:53.000  
3  2008-12-08 23:52:43.000  
4  2019-11-19 11:41:15.000